### PR TITLE
Init connectivity lib

### PR DESCRIPTION
**Pull request by LorenzE**
Add basic features for further connectivity features.
This is still WIP.

### CODE REVIEW

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 1]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+}
+
+
+//*************************************************************************************************************
+
+void BrainRTConnectivityDataTreeItem::plotNetwork(QSharedPointer<CONNECTIVITYLIB::Network> pNetworkData, const QVector3D& vecThreshold)
+{
+    //Prepare network visualization
+    QMatrix4x4 m;
+    Qt3DCore::QTransform* transform =  new Qt3DCore::QTransform();
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: really a pointer necessary? - smart pointer an option

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 2]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+    m.rotate(180, QVector3D(0.0f, 1.0f, 0.0f));
+    m.rotate(-90, QVector3D(1.0f, 0.0f, 0.0f));
+    transform->setMatrix(m);
+    m_pRenderable3DEntity->addComponent(transform);
+
+    QVector3D pos;
+    QSharedPointer<Qt3DCore::QEntity> sourceSphereEntity;
+    Qt3DExtras::QSphereMesh* sourceSphere;
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: never declare an uninitialized pointer, add = NULL;

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 3]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+    Qt3DExtras::QPhongMaterial* material;
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: add = NULL;

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 4]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+        pos.setY(lNetworkNodes.at(i)->getVert()(1));
+        pos.setZ(lNetworkNodes.at(i)->getVert()(2));
+
+        tMatVert(i,0) = pos.x();
+        tMatVert(i,1) = pos.y();
+        tMatVert(i,2) = pos.z();
+
+        sourceSphereEntity = QSharedPointer<Qt3DCore::QEntity>(new Qt3DCore::QEntity(m_pRenderable3DEntity));
+
+        sourceSphere = new Qt3DExtras::QSphereMesh();
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: is SourceSphere somewhere destructed?

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 5]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+        sourceSphere->setRadius(0.001f);
+        sourceSphereEntity->addComponent(sourceSphere);
+
+        Qt3DCore::QTransform* transform = new Qt3DCore::QTransform();
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: is transform somewhere destructed?

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp [hunk 6]
```diff
@@ -144,14 +142,161 @@ bool BrainRTConnectivityDataTreeItem::init(const MNEForwardSolution& tForwardSol
...
+        QMatrix4x4 m;
+        m.translate(pos);
+        transform->setMatrix(m);
+        sourceSphereEntity->addComponent(transform);
+
+        material = new Qt3DExtras::QPhongMaterial();
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: is material somewhere destructed?

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h [hunk 7]
```diff
@@ -150,7 +161,45 @@ class DISP3DNEWSHARED_EXPORT BrainRTConnectivityDataTreeItem : public AbstractTr
...
+    * Call this function whenever you want to calculate the indices/tris for a network.
+    *
+    * @param[in] pNetworkData     The network data.
+    * @param[in] vecThreshold     The threshold data.
+    */
+    void plotNetwork(QSharedPointer<CONNECTIVITYLIB::Network> pNetworkData, const QVector3D& vecThreshold);
+
+    bool                                        m_bIsInit;                      /**< The init flag. */
+
+    Qt3DCore::QEntity*                          m_pParentEntity;                /**< The parent 3D entity. */
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h]: please initialize

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h [hunk 8]
```diff
@@ -150,7 +161,45 @@ class DISP3DNEWSHARED_EXPORT BrainRTConnectivityDataTreeItem : public AbstractTr
...
+
+    Renderable3DEntity*                         m_pRenderable3DEntity;          /**< The renderable 3D entity. */
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h]: please initialize

File: MNE/disp3D/3DObjects/brain/brainsurfacesettreeitem.h [hunk 9]
```diff
@@ -218,8 +230,8 @@ class DISP3DNEWSHARED_EXPORT BrainSurfaceSetTreeItem : public AbstractTreeItem
     */
     void onColorInfoOriginChanged();
 
-    BrainRTSourceLocDataTreeItem*   m_pBrainRTSourceLocDataTreeItem;        /**< The rt data item of this hemisphere item. Multiple rt data item's can be added to this hemipshere item. */
-
+    BrainRTSourceLocDataTreeItem*       m_pBrainRTSourceLocDataTreeItem;        /**< The rt source loc data item of this hemisphere item. Multiple rt data item's can be added to this hemipshere item. */
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainsurfacesettreeitem.h]: please initialize

File: MNE/disp3D/3DObjects/brain/brainsurfacesettreeitem.h [hunk 10]
```diff
@@ -218,8 +230,8 @@ class DISP3DNEWSHARED_EXPORT BrainSurfaceSetTreeItem : public AbstractTreeItem
...
+    BrainRTConnectivityDataTreeItem*    m_pBrainRTConnectivityDataTreeItem;     /**< The rt connecntivity data item of this hemisphere item. Multiple rt data item's can be added to this hemipshere item. */
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainsurfacesettreeitem.h]: please initialize

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h (hunk 7 above)
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h]: and don't forget the delete

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h (hunk 8 above)
Comment (chdinh) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.h]: don't forget the delete

File: MNE/disp3D/3DObjects/data3Dtreemodel.cpp [hunk 11]
```diff
@@ -345,6 +346,68 @@ QList<BrainRTSourceLocDataTreeItem*> Data3DTreeModel::addData(const QString& sub
...
+            //Find already existing surface items and add the new data to the first search result
+            QList<QStandardItem*> itemList = pSubjectItem->findChildren(set);
+
+            if(!itemList.isEmpty() && (itemList.at(0)->type() == Data3DTreeModelItemTypes::SurfaceSetItem)) {
+                BrainSurfaceSetTreeItem* pSurfaceSetItem = dynamic_cast<BrainSurfaceSetTreeItem*>(itemList.at(0));
+                returnList.append(pSurfaceSetItem->addData(pNetworkData, m_pParentEntity));
+            } else {
+                BrainSurfaceSetTreeItem* pSurfaceSetItem = new BrainSurfaceSetTreeItem(Data3DTreeModelItemTypes::SurfaceSetItem, set);
+
+                QList<QStandardItem*> list;
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/data3Dtreemodel.cpp]: pointer? where is the garbage collection

File: MNE/disp3D/3DObjects/data3Dtreemodel.cpp [hunk 12]
```diff
@@ -345,6 +346,68 @@ QList<BrainRTSourceLocDataTreeItem*> Data3DTreeModel::addData(const QString& sub
...
+            SubjectTreeItem* pSubjectItem = dynamic_cast<SubjectTreeItem*>(itemSubjectList.at(i));
+
+            //Find already existing surface items and add the new data to the first search result
+            QList<QStandardItem*> itemList = pSubjectItem->findChildren(set);
+
+            if(!itemList.isEmpty() && (itemList.at(0)->type() == Data3DTreeModelItemTypes::SurfaceSetItem)) {
+                BrainSurfaceSetTreeItem* pSurfaceSetItem = dynamic_cast<BrainSurfaceSetTreeItem*>(itemList.at(0));
+                returnList.append(pSurfaceSetItem->addData(pNetworkData, m_pParentEntity));
+            } else {
+                BrainSurfaceSetTreeItem* pSurfaceSetItem = new BrainSurfaceSetTreeItem(Data3DTreeModelItemTypes::SurfaceSetItem, set);
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/data3Dtreemodel.cpp]: pointer - somewhere delete

File: MNE/disp3D/3DObjects/data3Dtreemodel.cpp [hunk 13]
```diff
@@ -345,6 +346,68 @@ QList<BrainRTSourceLocDataTreeItem*> Data3DTreeModel::addData(const QString& sub
 
 //*************************************************************************************************************
 
+QList<BrainRTConnectivityDataTreeItem*> Data3DTreeModel::addData(const QString& subject, const QString& set, Network::SPtr pNetworkData)
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/data3Dtreemodel.cpp]: list with pointer return??? - where is the delete?

File: MNE/disp3D/3DObjects/data3Dtreemodel.h [hunk 14]
```diff
@@ -195,6 +197,18 @@ class DISP3DNEWSHARED_EXPORT Data3DTreeModel : public QStandardItemModel
...
     /**
+    * Adds connectivity estimation data.
+    *
+    * @param[in] subject            The name of the subject.
+    * @param[in] set                The name of the surface set to which the actiavtion data is to be added.
+    * @param[in] pNetworkData       The connectivity data.
+    *
+    * @return                       Returns a list with the tree items which now hold the activation data. Use this list to update the data, i.e. during real time applications.
+    */
+    QList<BrainRTConnectivityDataTreeItem*> addData(const QString& subject, const QString& set, CONNECTIVITYLIB::Network::SPtr pNetworkData);
```
Comment (chdinh) [inline MNE/disp3D/3DObjects/data3Dtreemodel.h]: where is the delete of the pointers

File: MNE/connectivity/connectivitymeasures.cpp [hunk 15]
```diff
@@ -0,0 +1,279 @@
...
+//*************************************************************************************************************
+//=============================================================================================================
+// INCLUDES
+//=============================================================================================================
+
+#include "connectivitymeasures.h"
+#include "network/networknode.h"
+#include "network/networkedge.h"
+
+#include <iostream>
```
Comment (LorenzE) [inline MNE/connectivity/connectivitymeasures.cpp]: stl include section is deprecated and falls under "normal" include section

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp (hunk 1 above)
Comment (LorenzE) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: QEntity does not accpet smart/shared pointers. Note: If the Qt3DCore::QComponent has no parent, the Qt3DCore::QEntity will set itself as its parent thereby taking ownership of the component. This call should therefore be safe. What do you think?

File: MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp (hunk 4 above)
Comment (LorenzE) [inline MNE/disp3D/3DObjects/brain/brainrtconnectivitydatatreeitem.cpp]: sourceSphereEnt is th child of the QEntity. Once this one is destroyed QT will internally handle data management and deleting.

**Pull request merged**